src/Combinatorics/OrderedMultiIndex.jl

```diff
@@ -0,0 +1,182 @@
+    @assert all(k->i[k]<i[k+1], 1:length(i)-1) "indices must be strictly ordered"
+    return new{T}(i, n)
+  end
+end
+
+function ordered_multi_index(i::Vector{T}, n::T) where {T<:IntegerUnion} 
+  return OrderedMultiIndex(i, n)
+end
+
+# For an ordered multiindex i = (0 < i₁ < i₂ < … < iₚ ≤ n) this returns the vector (i₁,…,iₚ).
+indices(a::OrderedMultiIndex) = a.i
+
+# For an ordered multiindex i = (0 < i₁ < i₂ < … < iₚ ≤ n) this returns n.
+bound(a::OrderedMultiIndex) = a.n
+
+# For an ordered multiindex i = (0 < i₁ < i₂ < … < iₚ ≤ n) this returns p.
+length(a::OrderedMultiIndex) = length(a.i)
+
+# For an ordered multiindex i = (0 < i₁ < i₂ < … < iₚ ≤ n) this returns iₖ.
+getindex(a::OrderedMultiIndex, k::Int) = a.i[k]
+
+index_type(a::OrderedMultiIndex) = index_type(typeof(a))
+index_type(::Type{OrderedMultiIndex{T}}) where {T} = T
+
+# Internal function for "multiplication" of ordered multiindices.
+# 
+# The for i = (0 < i₁ < i₂ < … < iₚ ≤ n) and j = (0 < j₁ < j₂ < … < jᵣ ≤ n)
+# the result is a pair `(sign, a)` with `sign` either 0 in case that 
+# iₖ = jₗ for some k and l, or ±1 depending on the number of transpositions 
+# needed to put (i₁, …, iₚ, j₁, …, jᵣ) into a strictly increasing order 
+# to produce `a`.
+function _mult(a::OrderedMultiIndex{T}, b::OrderedMultiIndex{T}) where {T}
+  @assert bound(a) == bound(b) "multiindices must have the same bounds"
+
+  # in case of a double index return zero
+  any(x->(x in indices(b)), indices(a)) && return 0, a
+
+  p = length(a)
+  q = length(b)
+  result_indices = vcat(indices(a), indices(b))
+  sign = 1
+
+  # bubble sort result_indices and keep track of the sign
+  for k in p:-1:1
+    l = k
+    c = result_indices[l]
+    while l < p + q && c > result_indices[l+1]
+      result_indices[l] = result_indices[l+1]
+      sign = -sign
+      l = l+1
+    end
+    result_indices[l] = c
+  end
+  return sign, result_indices
+end
+
+# For two ordered multiindices i = (0 < i₁ < i₂ < … < iₚ ≤ n) 
+# and j = (0 < j₁ < j₂ < … < jᵣ ≤ n) this returns a pair `(sign, a)` 
+# with `sign` either 0 in case that iₖ = jₗ for some k and l, 
+# or ±1 depending on the number of transpositions needed to put 
+# (i₁, …, iₚ, j₁, …, jᵣ) into a strictly increasing order to produce `a`.
+function _wedge(a::OrderedMultiIndex{T}, b::OrderedMultiIndex{T}) where {T}
+  sign, ind = _mult(a, b)
+  iszero(sign) && return sign, a
+  return sign, OrderedMultiIndex(ind, bound(a))
+end
+
+function _wedge(a::Vector{T}) where {T <: OrderedMultiIndex}
+  isempty(a) && error("list must not be empty")
+  isone(length(a)) && return 1, first(a)
+  k = div(length(a), 2)
+  b = a[1:k]
+  c = a[k+1:end]
+  sign_b, ind_b = _wedge(b)
+  sign_c, ind_c = _wedge(c)
+  sign, ind = _wedge(ind_b, ind_c)
+  return sign * sign_b * sign_c, ind
+end
+
+function ==(a::OrderedMultiIndex{T}, b::OrderedMultiIndex{T}) where {T}
+  bound(a) == bound(b) || return false
+  return indices(a) == indices(b)
```

Comment:
```suggestion
  bound(a) == bound(b) && indices(a) == indices(b)
```
or
```suggestion
  return bound(a) == bound(b) && indices(a) == indices(b)
```
IMHO are easier to read and are (subjectively) more natural